lib/spack/spack/util/executable.py

```diff
@@ -7,53 +7,50 @@
 import subprocess
 import sys
 from pathlib import Path, PurePath
+from typing import Callable, Dict, Optional, Sequence, TextIO, Type, Union, overload
 
 import llnl.util.tty as tty
 
 import spack.error
-import spack.util.environment
+from spack.util.environment import EnvironmentModifications
 
 __all__ = ["Executable", "which", "which_string", "ProcessError"]
 
 
 class Executable:
     """Class representing a program that can be run on the command line."""
 
-    def __init__(self, name):
+    def __init__(self, name: str) -> None:
         file_path = str(Path(name))
         if sys.platform != "win32" and name.startswith("."):
             # pathlib strips the ./ from relative paths so it must be added back
             file_path = os.path.join(".", file_path)
-        self.exe = [file_path]
-
-        self.default_env = {}
 
-        self.default_envmod = spack.util.environment.EnvironmentModifications()
-        self.returncode = None
+        self.exe = [file_path]
+        self.default_env: Dict[str, str] = {}
+        self.default_envmod = EnvironmentModifications()
+        self.returncode = 0
```

Comment:
Can we add type annotations like this to the class body instead of the constructor in the future?  It is not as much of an issue here, as the constructor is simple, but as constructors evolve, it becomes increasingly hard to see what type everything is supposed to be. 

I don't see anything requiring it (though many places recommend it), and I don't know how we would lint for this, but I would like to move towards it.